### DISCUSSION TITLE
package-lock.json not detected in image scan, while Cargo.lock & composer.lock are

### DISCUSSION THREAD

**Discussion by john-d8r** [aquasecurity/trivy#4769] [Q&A] [answered]
### Description

While scanning an image from trivy's example https://github.com/aquasecurity/trivy-ci-test, noticed that `package-lock.json` is not detected, while `Cargo.lock` & `composer.lock` are detected.

I can understand that `package-lock.json` are not supported for image scans (as per trivy support matrix)

But I would like to understand the reason behind this decision and if there are any other work-around for this

### Desired Behavior

Trivy to detect `package-lock.json` while scanning image

### Actual Behavior

`package-lock.json` was not detected

### Reproduction Steps

```bash
1. clone https://github.com/aquasecurity/trivy-ci-test
2. `docker build . -t trivy-ci-test`
3. trivy image trivy-ci-test
```

### Target

Container Image

### Scanner

Vulnerability

### Output Format

None

### Mode

Standalone

### Debug Output

```bash
2023-07-04T13:37:47.478+0530    INFO    Detected OS: alpine
2023-07-04T13:37:47.478+0530    INFO    Detecting Alpine vulnerabilities...
2023-07-04T13:37:47.478+0530    DEBUG   alpine: os version: 3.7
2023-07-04T13:37:47.478+0530    DEBUG   alpine: package repository: 3.7
2023-07-04T13:37:47.478+0530    DEBUG   alpine: the number of packages: 58
2023-07-04T13:37:47.480+0530    INFO    Number of language-specific files: 2
2023-07-04T13:37:47.480+0530    INFO    Detecting composer vulnerabilities...
2023-07-04T13:37:47.480+0530    DEBUG   Detecting library vulnerabilities, type: composer, path: php-app/composer.lock
2023-07-04T13:37:47.480+0530    INFO    Detecting cargo vulnerabilities...
2023-07-04T13:37:47.480+0530    DEBUG   Detecting library vulnerabilities, type: cargo, path: rust-app/Cargo.lock
2023-07-04T13:37:47.489+0530    WARN    This OS version is no longer supported by the distribution: alpine 3.7.1
2023-07-04T13:37:47.489+0530    WARN    The vulnerability detection may be insufficient because security updates are not provided
```

### Operating System

WSL

### Version

```bash
Version: 0.42.1
Vulnerability DB:
  Version: 2
  UpdatedAt: 2023-07-04 06:09:14.346674696 +0000 UTC
  NextUpdate: 2023-07-04 12:09:14.346674396 +0000 UTC
  DownloadedAt: 2023-07-04 06:41:58.571723656 +0000 UTC
Java DB:
  Version: 1
  UpdatedAt: 2023-06-28 00:54:49.597173348 +0000 UTC
  NextUpdate: 2023-07-01 00:54:49.597172848 +0000 UTC
  DownloadedAt: 2023-06-28 03:04:49.960249058 +0000 UTC
```

### Checklist

- [X] Run `trivy image --reset`
- [X] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)

**Comment by knqyf263** [comment]
It is a bit complicated now. Please take a look at the doc to understand the behavior.
https://aquasecurity.github.io/trivy/v0.43/docs/scanner/vulnerability/language/

Trivy needs to scan `composer.lock` until Trivy supports PHP metadata of installed packages like `.gemspec` in Ruby.

**Comment by john-d8r** [comment]
Thanks for you reply @knqyf263 

> Trivy needs to scan `composer.lock` until Trivy supports PHP metadata of installed packages like `.gemspec` in Ruby.

Do you mean that once `Trivy supports PHP metadata of installed packages like .gemspec in Ruby.` support for composer.lock on image scans will be removed ?

Is there a reason for explicitly disabling lock file analyzers on image scan (for some languages) https://github.com/aquasecurity/trivy/blob/8e7fb7cc84bf010592115c6bfed602aff3410dec/pkg/commands/artifact/run.go#L168
I don't see trivy docs having explanation on that. (also it's not configurable ATM)

**Reply by knqyf263** [reply, marked as answer]
>Do you mean that once Trivy supports PHP metadata of installed packages like .gemspec in Ruby. support for composer.lock on image scans will be removed ?

Yes

>Is there a reason for explicitly disabling lock file analyzers on image scan (for some languages)

There could be unused lock files for testing or any other reasons in container images. Then, it leads to false detection. We've got many complaints about it.

> also it's not configurable ATM

What do you mean?

**Reply by john-d8r** [reply]
> There could be unused lock files for testing or any other reasons in container images. Then, it leads to false detection. We've got many complaints about it.

Ok understood

> What do you mean?

https://github.com/aquasecurity/trivy/blob/8e7fb7cc84bf010592115c6bfed602aff3410dec/pkg/flag/options.go#L108C1-L109C35
User can't control `DisabledAnalyzers` from cli at the moment, if they wants their lock files to be analyzed during image scans

**Reply by knqyf263** [reply]
It is the opposite, but there is a similar issue. We can discuss enabling analyzers as well.
https://github.com/aquasecurity/trivy/issues/3987

**Reply by john-d8r** [reply]
Thanks @knqyf263

**Reply by vikasdf** [reply]
@knqyf263 But with the current approach, we pick non-lock package manifest files (package.json, requirements.txt etc) even when lock file is available. These non-lock package manifests may not have the exact versions of the dependencies leading to errors in SBOM and associated vulnerability information. Also, we lose out on details like dependency tree etc. If some users left unused lock files in their image, shouldn't the proper fix be to advise them to remove the unused lock files? Also as John mentioned, this behavior should be configurable as there might be users who would prefer to use lockfiles when available.

**Reply by knqyf263** [reply]
You're confused with lock files and others. Let's rephrase them into the dependency list and the dependency itself. `requirements.txt` represents a list of dependencies. Image scanning will not pick up `requirements.txt` and instead use egg and wheel for python packages. As you said, `requirements.txt` may not include transitive dependencies, but Trivy can detect them from egg/whell.

**Reply by vikasdf** [reply]
I see. Thank you for clarifying that. I found a similar statement present [in the docs](https://aquasecurity.github.io/trivy/v0.43/docs/scanner/vulnerability/language/nodejs/#packagejson)
_Trivy searches for package.json files under node_modules and identifies installed packages. It only extracts package names, versions and licenses for those packages._

A side effect is that we noticed that dependency graph is not available for container scans (because package-lock.json is not scanned for container images). Any way to get the dependency tree (depends_on field) for nodejs packages for container image scans?

**Reply by knqyf263** [reply]
>Any way to get the dependency tree (depends_on field) for nodejs packages for container image scans?

There is no way at the moment. You need to scan `package-lock.json` with fs scanning.

**Reply by nahuel11500** [reply]
Hello, any update on this ? Would it be possible to add a flag to enable package-lock.json scanning in image like that ? : 

`trivy image --enable-analyzer`

I would be happy to provide a PR if it's okay with you.
